src/main/java/org/pac4j/vertx/handler/impl/RequiresAuthenticationHandler.java

```diff
@@ -0,0 +1,95 @@
+package org.pac4j.vertx.handler.impl;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.auth.User;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.impl.AuthHandlerImpl;
+import org.pac4j.core.client.Client;
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.config.Config;
+import org.pac4j.core.profile.ProfileManager;
+import org.pac4j.core.profile.UserProfile;
+import org.pac4j.vertx.VertxWebContext;
+import org.pac4j.vertx.auth.Pac4jAuthProvider;
+import org.pac4j.vertx.auth.Pac4jUser;
+import org.pac4j.vertx.flow.AuthenticationFlow;
+
+import java.util.function.Consumer;
+
+/**
+ * @author jez
+ */
+public abstract class RequiresAuthenticationHandler extends AuthHandlerImpl {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RequiresAuthenticationHandler.class);
+  protected final Config config;
+  protected final String clientName;
+  protected final Vertx vertx;
+  private final boolean allowDynamicClientSelection;
+
+  public RequiresAuthenticationHandler(final Vertx vertx, final Config config, final Pac4jAuthProvider authProvider,
+                                       final Pac4jAuthHandlerOptions options) {
+    super(authProvider);
+    clientName = options.clientName();
+    allowDynamicClientSelection = options.allowDynamicClientSelection();
+    this.vertx = vertx;
+    this.config = config;
+  }
+
+  // Port of Pac4J auth to a handler in vert.x 3.
+  @Override
+  public void handle(RoutingContext routingContext) {
+
+    final User user = routingContext.user();
+    if (user != null) {
+      // Already logged in, just authorise
+      authorise(user, routingContext);
+    } else {
+      final VertxWebContext webContext = new VertxWebContext(routingContext);
+      final Client client = findClient(webContext, this.clientName, allowDynamicClientSelection);
+      final ProfileManager profileManager = new ProfileManager(webContext);
+      final AuthenticationFlow authenticationFlow = getAuthenticationFlow(client);
```

Comment:
In the DirectAuthenticationFlow::initiate method (in the case of direct, initiate performs the full authentication). The handler passed in then deals with the profile returned (which can be null, see other issue for the fact that I need to ensure that happens).